Do not close container from reconnectOnError even if canRetry is false

markfields opened the pull request:
Fixes #8570

There should be no behavior change with this PR.  But the flow/expectations are much clearer.

Previously, 2/3 codepaths entering `reconnectOnError` would stomp over `canRetry` with `true`, to avoid closing the container due to hiccup on the socket (assumption being, any fatal error state will cause container to close on reconnect attempt).  Now, rather than overwriting `canRetry`, I've let that stay as-is and updated `reconnectOnError` to simply log if `canRetry` is false.

The third codepath entering `reconnectOnError` (nack) has a case where `canRetry` is false, so in that case rather than leaning on `reconnectOnError` to close the container, I check `canRetry` first and close right away (then `reconnectOnError` will return early).

Note: My previous analysis that stated that `documentDeltaConnection`'s `disconnect` event could lead to this code path if raised during `odspDocumentDeltaConnection.initialize` was incorrect - at that point in time, `connectionManager` hasn't registered event listeners, of course, since the connection object is never successfully created/returned.

With this in mind (and given the refactor for the non-retryable nack), I don't actually expect the new log event `"reconnectingDespiteFatalError"` to ever be hit.  And even if it is, it's an expected/supported code path so I'm logging it as a generic event, not an error event.

markfields committed:
Don't force canRetry to be true when reconnecting.

Also some adjustments to errors in connectionManager

markfields committed:
Another adjusted errorType

Trying to avoid using driver errors not from the driver.
Exception seems to be connectionManager.

markfields committed:
Add IAnyDriverError type in shared driver layer

markfields committed:
Use IAnyDriverError in DeltaManager/ConnectionManager

Event handlers are typed with IAnyDriverError
even though the interface still says any.
Will move the new type to driver-definitions later.

markfields committed:
Remove unrelated/unnecessary changes

markfields committed:
build fix

andre4i commented on packages/loader/container-loader/src/connectionManager.ts:
```diff
@@ -743,8 +740,8 @@ export class ConnectionManager implements IConnectionManager {
-        const delayMs = error !== undefined ? getRetryDelayFromError(error) : undefined;
-        if (delayMs !== undefined) {
+        const delayMs = getRetryDelayFromError(error);
+        if (error !== undefined && delayMs !== undefined) {
```
is `error !== undefined` redundant here?

markfields replied:
Correct.  But without it this way, TypeScript can't figure out the type narrowing and complains that `error` could be `undefined` on the next line.  If you have a stronger opinion, I can do `error!` on the next line with a lint error suppression, but I prefer to avoid those myself.

markfields commented:
The test failures are important, I'm taking a look.  I understand that this change will result in container closing immediately, rather than from `connectCore` where it awaits the promise returned from `initialize`.  That seems to be breaking some assumptions in these tests.

andre4i commented on packages/loader/driver-utils/src/network.ts:
```diff
@@ -29,6 +29,18 @@ export function isOnline(): OnlineStatus {
+    readonly errorType: string;
+    readonly message: string;
+    canRetry: boolean;
+    online?: string;
```
some comments here would be nice :) `online` doesn't say much about what it means.

vladsud replied:
I'm not sure "Any" has a value.

markfields replied:
You'd prefer `IDriverError`?  I'm trying to articulate that this can be an error from the base/utils driver code, or from a specialized (e.g. ODSP) driver layer.

vladsud replied:
I get it, but same logic applies to any driver interfaces, and we do not put "any" anywhere.
Some other places use "IGenericDriverError". Not sure it's better.
You can also go with IDocumentServiceError - this seems like more concrete (one implication - it assumes it's used only by instance, not by driver factory).

I think the core issue here is that it's not part of interface, i.e., we need to move toward drivers returning errors, not throwing errors. Then it becomes a bit easier to name (and "any" would make even less sense there).

markfields replied:
> we need to move toward drivers returning errors, not throwing errors

Yes that is the direction.  Trying to make tactical fixes with that future in mind.  This type will be replaced with a proper type (with strongly-typed errorType) that can account for driver-specific errors in a careful way.

markfields replied:
Leaving it named as-is with a comment explaining a bit better its shape and purpose

andre4i commented on packages/loader/container-loader/src/connectionManager.ts:
```diff
@@ -726,7 +723,7 @@ export class ConnectionManager implements IConnectionManager {
         this.disconnectFromDeltaStream(disconnectMessage);
 
-        const canRetry = error !== undefined ? canRetryOnError(error) : true;
+        const canRetry = (error === undefined) || canRetryOnError(error);
```
Not saying this is wrong, considering the code being logically equivalent. But I'm curious why we'd retry on an undefined error. It seems like this method might be used for a generic reconnect without an error, which is slightly against the name of the method. Nothing actionable right now but something to keep in mind in case of a larger refactoring.

andre4i commented:
> The test failures are important, I'm taking a look. I understand that this change will result in container closing immediately, rather than from connectCore where it awaits the promise returned from initialize. That seems to be breaking some assumptions in these tests.

Please see if you can add new tests to lock in the new behavior explicitly. I mean in case the existing tests aren't explicit about why they fail..

markfields committed:
More narrow typings (replacing any)

markfields commented:
As discussed in the issue, I misunderstood the semantics here.  Will update the PR to restore the old behavior but make it more explicit/clear/consistent.

markfields committed:
reconnectOnError no longer closes the container if canRetry is false

And the one place that was relying on that behavior
now closes the container itself rather than calling into it

markfields committed:
Merge branch 'main' into canretry-preserved

markfields commented:
@andre4i, @vladsud  -- Ready for review again.  I'm updating the title/original comment to explain the new approach, which is to leave canRetry as-is, but don't close if it's false.

markfields committed:
Revert overzealous typing

It will be so in practice, but not provable in the code today
Will follow up.

vladsud commented on packages/loader/container-loader/src/connectionManager.ts:
```diff
@@ -68,6 +67,12 @@ const DefaultChunkSize = 16 * 1024;
 
 function getNackReconnectInfo(nackContent: INackContent) {
+    // check message.content for Back-compat with old service.
+    if (nackContent === undefined) {
```
not sure what "old service" means in this context. Is this some kind of new thing?
If it's an old thing, how did code work before?

markfields replied:
I just copied this block from below.  I bet we could remove it in a separate PR once we identified the back-compat scenario/timeline

vladsud replied:
Got it, I did not see the code where it moved from :)

vladsud commented on packages/loader/container-loader/src/connectionManager.ts:
```diff
@@ -726,12 +724,7 @@ export class ConnectionManager implements IConnectionManager {
-        // If reconnection is not an option, close the DeltaManager
-        if (!canRetry) {
-            this.props.closeHandler(normalizeError(error, { props: fatalConnectErrorProp }));
-        } else if (this.reconnectMode === ReconnectMode.Never) {
```
not sure why canRetry handling is removed. Is this intentional?
I believe nackHandler() workflow can result in non-retryable error. I.e. getNackReconnectInfo() & nackContent.code !== 403.

markfields replied:
Please read PR description

vladsud replied:
Thanks! I'd suggest for nackHandler() to not even call this function in case of canRetry - false, and just add assert(canRetry, "all code paths reach here with retryable error") and remove any handling or events below - it feels clearly RE enforcing invariants and capturing any changes in workflows sooner.

markfields replied:
Might not be true though.  See error handler in documentDeltaConnection

markfields replied:
```
const err = this.createErrorObject("error", error, error !== "Invalid namespace");
```

vladsud replied:
If my memory serves right, this case will result in failed connection, not disconnect. It happens before we get connect_document_success. I did not retraced code again, but I believe all reconnection flows were always passing canRetry = true except for nack path, so even if this error can result in disconnect, we are not handling it correctly.

markfields replied:
I think you're right. I want to start with the log and see if it bears out, I can follow up later and switch to assert. Not going to open an issue, I don't think it is critical to ensure it happens.

markfields committed:
build fix

markfields committed:
build fix

markfields committed:
Update comment

markfields committed:
Add reconnectOnError unit test

markfields committed:
build fix

markfields committed:
Remove spurious .only

markfields commented:
@andre4i, @vladsud  -- I added tests and made some trivial changes, feel free to leave further comments, I am opening a follow-up PR with some stuff anyway.
